pkg/controller/vitessshard/reconcile_rollout.go

```diff
@@ -0,0 +1,137 @@
+			planetscalev2.KeyspaceLabel:  vts.Labels[planetscalev2.KeyspaceLabel],
+			planetscalev2.ShardLabel:     vts.Spec.KeyRange.SafeName(),
+		}).AsSelector(),
+	}
+
+	if err := r.client.List(ctx, listOpts, podList); err != nil {
+		return resultBuilder.Error(err)
+	}
+
+	// Safety checks and rolling updates must be deterministically ordered, so we access and sort pods by tablet alias.
+	tabletPods := make(map[string]*v1.Pod)
+	for i := range podList.Items {
+		pod := &podList.Items[i]
+		tabletAlias := vttablet.AliasFromPod(pod)
+		tabletKey := topoproto.TabletAliasString(&tabletAlias)
+		tabletPods[tabletKey] = pod
+	}
+
+	tabletKeys := make([]string, 0, len(vts.Status.Tablets))
+	for key := range vts.Status.Tablets {
+		tabletKeys = append(tabletKeys, key)
+	}
+	sort.Strings(tabletKeys)
+
+	for _, tabletKey := range tabletKeys {
+		tablet := vts.Status.Tablets[tabletKey]
+		if tablet.Available != v1.ConditionTrue {
+			// If any tablets are unhealthy, we should bail and not perform a rolling restart.
+			r.recorder.Eventf(vts, v1.EventTypeNormal, "RolloutPaused", "Waiting for tablet %v to be Available.", tabletKey)
+			return resultBuilder.Result()
+		}
+
+		pod, ok := tabletPods[tabletKey]
+		if !ok {
+			r.recorder.Eventf(vts, v1.EventTypeNormal, "RolloutPaused", "Waiting for desired tablet %v to be created.", tabletKey)
+			return resultBuilder.Result()
+		}
+
+		if rollout.Released(pod) {
+			// If any tablet has already been released, we should wait until it is finished to release another one.
+			r.recorder.Eventf(vts, v1.EventTypeNormal, "RolloutPaused", "Waiting for tablet %v to finish release.", tabletKey)
+			return resultBuilder.Result()
+		}
+	}
+
+	// Retrieve tablet pod to be released during this reconcile.
+	tabletKey, pod := getNextScheduledTablet(tabletKeys, tabletPods)
+	if tabletKey == "" {
+		// If we have no more scheduled tablets, uncascade the shard.
+		if err := r.uncascadeShard(ctx, vts); err != nil {
+			r.recorder.Eventf(vts, v1.EventTypeWarning, "UncascadeFailed", "Client failed to mark shard as uncascading.")
+			return resultBuilder.Error(err)
+		}
+
+		r.recorder.Eventf(vts, v1.EventTypeNormal, "CascadeComplete", "Cascading changes have been successfully applied to shard.")
+		return resultBuilder.Result()
+	}
+
+	masterEligibleTablets := vts.Spec.MasterEligibleTabletCount()
+	deletePod := false
+	tabletType := pod.Labels[planetscalev2.TabletTypeLabel]
+	// These two conditions guarantee that the tablet is a lone master.
+	if masterEligibleTablets == 1 && tabletType == string(planetscalev2.ReplicaPoolType) {
+		// If we have a lone master, we must delete it since reparenting is impossible.
+		deletePod = true
+	}
+
+	if err := r.releaseTabletPod(ctx, pod, deletePod); err != nil {
+		r.recorder.Eventf(vts, v1.EventTypeWarning, "RollingRestartBlocked", "release of Pod %v (tablet %v) failed: %v", pod.Name, tabletKey, err)
+		return resultBuilder.Error(err)
```

Comment:
At this point, now that we aren't in a loop, anymore, so I think preferring less exit points will be clearer. We will return `resultBuilder.Result()` in a moment anyways, so we can just do `resultBuilder.Error(err)` with no return here and then return (as we already will) at the final exit on line 105. Calling `resultBuilder.Result()` on line 105 will carry with it any errors from `resultBuilder.Error(err)`.